src/torrent.cpp

```diff
@@ -5291,12 +5291,20 @@ namespace libtorrent
 		TORRENT_ASSERT(std::count(m_peers_to_disconnect.begin()
 			, m_peers_to_disconnect.end(), p) == 0);
 
-		std::weak_ptr<torrent> weak_t = shared_from_this();
-		m_peers_to_disconnect.push_back(p);
-		m_deferred_disconnect.post(m_ses.get_io_service(), [=]() {
-			std::shared_ptr<torrent> t = weak_t.lock();
-			if (t) t->on_remove_peers();
-		});
+		// we only schedule the peer for actual removal if in fact
+		// it is attached to this torrent, since this method can
+		// be called from the attach_peer path and fail to do so
+		// because of too many connections
+		if (p->associated_torrent().lock().get() == this)
```

Comment:
indeed, `p` is part of the `m_connections`, but when coming via `peer_connection::attach_to_torrent`->`torrent::attach_peer`-->`maybe_replace_peer is true`, and due to ranking call `remove_peer(p);`, the `peer_connection* p` is left without associated torrent. Then `session_impl::tick` could clean it, and we end up with an invalid pointer in `m_connections`